File: pkg/service/recap_service.go
package service

import (
	"fmt"
	"time"

	"teaching-management/pkg/model"
	"teaching-management/pkg/repository"

	"github.com/google/uuid"
)

type RecapService struct {
	sessionRepo repository.SessionRepository
	studentRepo repository.StudentRepository
}

func NewRecapService(sessionRepo repository.SessionRepository, studentRepo repository.StudentRepository) *RecapService {
	return &RecapService{
		sessionRepo: sessionRepo,
		studentRepo: studentRepo,
	}
}

func (s *RecapService) GetMonthlyRecap(userID uuid.UUID, year, month int) (*model.MonthlyRecapResponse, error) {
	startDate := time.Date(year, time.Month(month), 1, 0, 0, 0, 0, time.UTC)
	endDate := startDate.AddDate(0, 1, -1)

	startDateStr := startDate.Format("2006-01-02")
	endDateStr := endDate.Format("2006-01-02")

	sessions, err := s.sessionRepo.GetAll(userID, startDateStr, endDateStr, nil)
	if err != nil {
		return nil, err
	}

	students, err := s.studentRepo.GetAll(userID)
	if err != nil {
		return nil, err
	}

	recapMap := make(map[uuid.UUID]model.StudentRecap)
	for _, st := range students {
		recapMap[st.ID] = model.StudentRecap{
			StudentID:        st.ID,
			StudentName:      st.FullName,
			FeeModel:         st.FeeModel,
			FeeAmount:        st.FeeAmount,
			CompletedCount:   0,
			CancelledCount:   0,
			RescheduledCount: 0,
			TotalEarnings:    0.0,
		}
	}

	var completed, cancelled, rescheduled int

	for _, se := range sessions {
		r, exists := recapMap[se.StudentID]
		if !exists {
			r = model.StudentRecap{
				StudentID:        se.StudentID,
				StudentName:      se.StudentName,
				CompletedCount:   0,
				CancelledCount:   0,
				RescheduledCount: 0,
				TotalEarnings:    0.0,
			}
		}

		switch se.Status {
		case "completed":
			r.CompletedCount++
			completed++
			r.TotalEarnings += se.FeeCalculated
		case "cancelled":
			r.CancelledCount++
			cancelled++
		case "rescheduled":
			r.RescheduledCount++
			rescheduled++
			r.TotalEarnings += se.FeeCalculated
		}

		recapMap[se.StudentID] = r
	}

	var studentRecaps = []model.StudentRecap{}
	var totalEarnings float64

	for _, r := range recapMap {
		if r.FeeModel == "monthly" {
			totalSessions := r.CompletedCount + r.CancelledCount + r.RescheduledCount
			if totalSessions > 0 {
				r.TotalEarnings = r.FeeAmount
			} else {
				r.TotalEarnings = 0.0
			}
		}
		totalEarnings += r.TotalEarnings
		studentRecaps = append(studentRecaps, r)
	}

	return &model.MonthlyRecapResponse{
		Year:                year,
		Month:               month,
		TotalSessions:       len(sessions),
		CompletedSessions:   completed,
		CancelledSessions:   cancelled,
		RescheduledSessions: rescheduled,
		TotalEarnings:       totalEarnings,
		Students:            studentRecaps,
	}, nil
}

func (s *RecapService) GetMonthlyTrend(userID uuid.UUID, limit int) ([]model.MonthlyTrendItem, error) {
	if limit <= 0 {
		limit = 6
	}

	now := time.Now()
	var trend = []model.MonthlyTrendItem{}

	for i := limit - 1; i >= 0; i-- {
		t := now.AddDate(0, -i, 0)
		year := t.Year()
		month := int(t.Month())

		recap, err := s.GetMonthlyRecap(userID, year, month)
		if err != nil {
			return nil, err
		}

		monthLabels := []string{"", "Jan", "Feb", "Mar", "Apr", "Mei", "Jun", "Jul", "Agt", "Sep", "Okt", "Nov", "Des"}
		label := fmt.Sprintf("%s %d", monthLabels[month], year)

		trend = append(trend, model.MonthlyTrendItem{
			Label:    label,
			Earnings: recap.TotalEarnings,
		})
	}

	return trend, nil
}

File: src/services/recap.service.ts
import {
  SessionRepository,
  StudentRepository
} from '../repositories/interfaces.js';
import {
  MonthlyRecapResponse,
  MonthlyTrendItem,
  StudentRecap
} from '../types/index.js';

export class RecapService {
  constructor(
    private sessionRepo: SessionRepository,
    private studentRepo: StudentRepository
  ) {}

  async getMonthlyRecap(
    userId: string,
    year: number,
    month: number
  ): Promise<MonthlyRecapResponse> {
    const start = new Date(Date.UTC(year, month - 1, 1));
    const end = new Date(Date.UTC(year, month, 0)); // last day of month

    const startDateStr = start.toISOString().slice(0, 10);
    const endDateStr = end.toISOString().slice(0, 10);

    const [sessions, students] = await Promise.all([
      this.sessionRepo.getAll(userId, startDateStr, endDateStr),
      this.studentRepo.getAll(userId)
    ]);

    const recapMap = new Map<string, StudentRecap>();
    for (const st of students) {
      recapMap.set(st.id, {
        student_id: st.id,
        student_name: st.full_name,
        fee_model: st.fee_model,
        fee_amount: Number(st.fee_amount || 0),
        completed_count: 0,
        cancelled_count: 0,
        rescheduled_count: 0,
        total_earnings: 0.0
      });
    }

    let completed = 0;
    let cancelled = 0;
    let rescheduled = 0;

    for (const se of sessions) {
      let r = recapMap.get(se.student_id);
      if (!r) {
        r = {
          student_id: se.student_id,
          student_name: se.student_name,
          fee_model: 'per_session',
          fee_amount: 0,
          completed_count: 0,
          cancelled_count: 0,
          rescheduled_count: 0,
          total_earnings: 0.0
        };
        recapMap.set(se.student_id, r);
      }

      switch (se.status) {
        case 'completed':
          r.completed_count++;
          completed++;
          r.total_earnings += Number(se.fee_calculated || 0);
          break;
        case 'cancelled':
          r.cancelled_count++;
          cancelled++;
          break;
        case 'rescheduled':
          r.rescheduled_count++;
          rescheduled++;
          r.total_earnings += Number(se.fee_calculated || 0);
          break;
      }
    }

    const studentRecaps: StudentRecap[] = [];
    let totalEarnings = 0.0;

    for (const r of recapMap.values()) {
      if (r.fee_model === 'monthly') {
        const totalStudentSessions =
          r.completed_count + r.cancelled_count + r.rescheduled_count;
        if (totalStudentSessions > 0) {
          r.total_earnings = Number(r.fee_amount || 0);
        } else {
          r.total_earnings = 0.0;
        }
      }
      totalEarnings += r.total_earnings;
      studentRecaps.push(r);
    }

    return {
      year,
      month,
      total_sessions: sessions.length,
      completed_sessions: completed,
      cancelled_sessions: cancelled,
      rescheduled_sessions: rescheduled,
      total_earnings: totalEarnings,
      students: studentRecaps
    };
  }

  async getMonthlyTrend(userId: string, limit = 6): Promise<MonthlyTrendItem[]> {
    if (limit <= 0) limit = 6;

    const now = new Date();
    const trend: MonthlyTrendItem[] = [];

    const monthLabels = [
      '',
      'Jan',
      'Feb',
      'Mar',
      'Apr',
      'Mei',
      'Jun',
      'Jul',
      'Agt',
      'Sep',
      'Okt',
      'Nov',
      'Des'
    ];

    for (let i = limit - 1; i >= 0; i--) {
      const d = new Date(now.getFullYear(), now.getMonth() - i, 1);
      const year = d.getFullYear();
      const month = d.getMonth() + 1;

      const recap = await this.getMonthlyRecap(userId, year, month);
      const label = `${monthLabels[month]} ${year}`;

      trend.push({
        label,
        earnings: recap.total_earnings
      });
    }

    return trend;
  }
}
